specs/AppLifecycle/Restart/restartApi.markdown

```diff
@@ -0,0 +1,259 @@
+
+<table>
+<colgroup>
+<col style="width: 41%" />
+<col style="width: 58%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>Existing API</th>
+<th>Details</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td><p>HRESULT <strong>RegisterApplicationRestart</strong>(PCWSTR pwzCommandline, DWORD dwFlags);</p></td>
+<td><p>The pwzCommandline parameter is passed into the new instance as its command-line arguments string. The dwFlags can be 0 or one or more of the (flags) values:</p>
+<ul>
+<li>
+<p>1: RESTART_NO_CRASH - Do not restart the process if it terminates due to an unhandled exception.</p>
+</li>
+<li>
+<p>2: RESTART_NO_HANG - Do not restart the process if it terminates due to the application not responding.</p>
+</li>
+<li>
+<p>4: RESTART_NO_PATCH - Do not restart the process if it terminates due to the installation of an OS update.</p>
+</li>
+<li>
+<p>8: RESTART_NO_REBOOT - Do not restart the process if the computer is restarted as the result of an OS update.</p>
+</li>
+</ul></td>
+</tr>
+<tr class="even">
+<td>HRESULT <strong>UnregisterApplicationRestart</strong>();</td>
+<td>Unregisters the app for restart, where the app had previously called RegisterApplicationRestart.</td>
+</tr>
+<tr class="odd">
+<td>public static IAsyncOperation&lt;AppRestartFailureReason&gt; <strong>RequestRestartAsync</strong>(<br />
+string launchArguments);</td>
+<td><p>Static function in CoreApplication. Requests immediate termination and restart.</p>
+<ul>
+<li>
+<p>The launchArguments string is passed into the new instance as the Arguments property on LaunchActivatedEventArgs.</p>
+</li>
+<li>
+<p>Also sets LaunchActivatedEventArgs.PreviousExecutionState to Terminated.</p>
+</li>
+</ul></td>
+</tr>
+<tr class="even">
+<td>public enum <strong>AppRestartFailureReason</strong></td>
+<td><ul>
+<li>
+<p>0: RestartPending - a restart is already in progress.</p>
+</li>
+<li>
+<p>1: NotInForeground - an app must be visible and in the foreground when it calls the restart API.</p>
+</li>
+<li>
+<p>2: InvalidUser - could not restart for the specified user.</p>
+</li>
+<li>
+<p>3: Other - unspecified failure.</p>
+</li>
+</ul></td>
+</tr>
+<tr class="odd">
+<td>class <strong>LaunchActivatedEventArg</strong></td>
+<td><p>Relevant Properties:</p>
+<ul>
+<li><p>Arguments: gets the arguments that are passed to the app during its launch activation.</p></li>
+<li><p>Kind: Gets the reason that this app is being activated. This is of type enum ActivationKind</p></li>
+</ul>
+<p>PreviousExecutionState: Gets the execution state of the app before this activation.</p></td>
+</tr>
+</tbody>
+</table>
+
+## New API Details
+
+### RequestRestartNow
+
+<table>
+<colgroup>
+<col style="width: 41%" />
+<col style="width: 53%" />
+<col style="width: 5%" />
+</colgroup>
+<thead>
+<tr class="header">
+<th>New API </th>
+<th>Description</th>
+</tr>
+</thead>
+<tbody>
+<tr class="odd">
+<td>static AppRestartFailureReason <strong>RequestRestartNow</strong>(string restartArguments)</td>
```

Comment:
+1 to removing "request". Actually I think `Restart()` might be enough.

Reply:
This is something we've debated in the past. We chose to go this path for mainly 2 reasons:
1. Technically a restart is being "requested", it could fail. It may not actually complete successfully and there may be another restart already in progress/pending.
2. For parity/consistency with the UWP RequestRestartAsync

Reply:
DISCUSSION: This is on the AppInstance itself, so it's to restart the instance with new args. The Request- prefix is to align better with the existing Windows.ApplicationModel.CoreApplication method, which needs permission from the user.  We'll do some investigation to figure out what the Request- operation actually did in the OS.  If this API is used in a UWP it will need to be on a UI thread.

RECOMMEND: Try calling it Restart, unless we find that the Request was integral to the original design.